File: MapboxVelocity/MapboxVelocity/Features/Map/TripModel.swift
//
//  TripModel.swift
//  MapboxVelocity
//
//  Created by Shaikat on 4/26/26.
//

import CoreLocation

struct TripLocation {
    let coordinate: CLLocationCoordinate2D
    let address: String
}

enum VehicleType: CaseIterable, Hashable {
    case bike
    case car
    case scooter

    var title: String {
        switch self {
        case .bike:    "Bike"
        case .car:     "Car"
        case .scooter: "Scooter"
        }
    }

    var basePrice: Double {
        switch self {
        case .bike:    77.00
        case .car:     99.00
        case .scooter: 66.00
        }
    }

    var icon: String {
        switch self {
        case .bike:    "bicycle"
        case .car:     "car.fill"
        case .scooter: "scooter"
        }
    }

    // Degrees to subtract from bearing so the SF symbol faces north at 0°
    var bearingOffset: Double {
        switch self {
        case .car:           180  // car.fill faces left (west)
        case .bike, .scooter: 90  // bicycle/scooter face right (east)
        }
    }
}

enum PaymentType {
    case cash
    case mobilePay
}

enum TripState: Equatable {
    case idle
    case pickupSet
    case fetchingRoute
    case routeReady
    case riding
    case rideEnded
    case error(String)
}

struct Trip {
    var pickup: TripLocation?
    var destination: TripLocation?
    var vehicle: VehicleType
    var payment: PaymentType
    var distanceMeters: Double
    var durationSeconds: Double
    var discount: Double

    var finalPrice: Double { vehicle.basePrice - discount }

    static var empty: Trip {
        Trip(pickup: nil, destination: nil, vehicle: .car,
             payment: .cash, distanceMeters: 0, durationSeconds: 0, discount: 0)
    }
}

File: MapboxVelocityAndroid/app/src/main/java/com/mapboxvelocity/Features/Map/TripModel.kt
package com.mapboxvelocity.Features.Map

import com.mapbox.geojson.Point

data class TripLocation(
    val point: Point,
    val address: String
)

enum class VehicleType {
    BIKE, CAR, SCOOTER;

    val title: String get() = when (this) {
        BIKE    -> "Bike"
        CAR     -> "Car"
        SCOOTER -> "Scooter"
    }

    val basePrice: Double get() = when (this) {
        BIKE    -> 77.00
        CAR     -> 99.00
        SCOOTER -> 66.00
    }

    // Emoji icons — Android has no SF Symbols; mirrors iOS icon intent
    val emoji: String get() = when (this) {
        BIKE    -> "🚲"
        CAR     -> "🚗"
        SCOOTER -> "🛵"
    }

    val bearingOffset: Double get() = when (this) {
        CAR             -> 180.0
        BIKE, SCOOTER   -> 90.0
    }
}

enum class PaymentType {
    CASH, MOBILE_PAY;

    val title: String get() = when (this) {
        CASH        -> "Cash"
        MOBILE_PAY  -> "Mobile Pay"
    }
}

sealed class TripState {
    object Idle         : TripState()
    object PickupSet    : TripState()
    object FetchingRoute: TripState()
    object RouteReady   : TripState()
    object Riding       : TripState()
    object RideEnded    : TripState()
    data class Error(val message: String) : TripState()

    override fun equals(other: Any?): Boolean = this === other || this::class == other?.let { it::class }
    override fun hashCode(): Int = this::class.hashCode()
}

data class Trip(
    val pickup: TripLocation?       = null,
    val destination: TripLocation?  = null,
    val vehicle: VehicleType        = VehicleType.CAR,
    val payment: PaymentType        = PaymentType.CASH,
    val distanceMeters: Double      = 0.0,
    val durationSeconds: Double     = 0.0,
    val discount: Double            = 0.0
) {
    val finalPrice: Double get() = vehicle.basePrice - discount

    companion object {
        val empty get() = Trip()
    }
}
